r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -131,6 +132,56 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(res)
     }
 
+    /// Construct a NonNativeFieldMulResult gadget from a NonNativeFieldGadget
+    pub(crate) fn from_non_native_field_gadget<CS: ConstraintSystemAbstract<ConstraintF>>(mut cs: CS, input_gagdet: &NonNativeFieldGadget<SimulationF, ConstraintF>,
+        params: &NonNativeFieldParams) -> Self {
+
+        debug_assert!(
+            input_gagdet.check(),
+            "convert to mul result gagdet from non native gadget: check() failed on input gadget"
+        );
+
+        let mut limbs = vec![FpGadget::<ConstraintF>::from_value(&mut cs, &ConstraintF::zero()); params.num_limbs-1];
+        limbs.extend_from_slice(&input_gagdet.limbs[..]);
+        // set num_add to ceil(input_gadget.num_add/2^bits_per_limb) as a limb of
+        // NonNativeFieldMulResultGadget has 2*bits_per_limb for each limb
+        let num_add_over_normal_form = &input_gagdet.num_of_additions_over_normal_form/BigUint::from(2usize).pow(params.bits_per_limb as u32) + BigUint::from(if &input_gagdet.num_of_additions_over_normal_form % BigUint::from(2usize).pow(params.bits_per_limb as u32) != BigUint::from(0usize) { 1usize } else {0usize});
```

Comment:
This can be made a bit more precise:
If _n_ is the number of additions over normal form of the initial gadget "_x_", then we have that
_x < (n+1) * (2^bpl - 1)_
which is equivalent to
_x < { (n+1) / (2^bpl - 1) } * (2^bpl - 1)^2_
It follows, called _n'_ the new number of additions over normal form, _n'_ = ceil _((n+1) / (2^bpl - 1)  - 1)_
In practice there will be no difference, since this will be always 0.

Reply:
You are right about the n+1, I forgot it, thanks. But it is not clear why we should consider 2^bpl-1 as an upper bound for the limb rather than 2^bpl.

Reply:
I am not sure about the validity of the implemented formula. In terms of the limb bound factors `num_adds + 1`, shouldn't we have 
```
   (num_adds + 1) = (num_adds(product) + 1) *  2^bits_per_limb,
```
and hence ``num_adds(product) = Floor[(num_adds + 1)/2^bits_per_limb] - 1``?

Reply:
> You are right about the n+1, I forgot it, thanks. But it is not clear why we should consider 2^bpl-1 as an upper bound for the limb rather than 2^bpl.

just because 2^bpl - 1 = (1...1)_2, but it does not make a big difference (in this case no difference in practice).

Reply:
> I am not sure about the validity of the implemented formula. In terms of the limb bound factors `num_adds + 1`, shouldn't we have
> 
> ```
>    (num_adds + 1) = (num_adds(product) + 1) *  2^bits_per_limb,
> ```
> 
> and hence `num_adds(product) = Floor[(num_adds + 1)/2^bits_per_limb] - 1`?

num_adds(product) = Floor[(num_adds + 1)/2^bits_per_limb] - 1

This could be a negative number (in practice it will be -1), we need to put ceil instead of floor.

Reply:
I would still use 2^bpl since, at least for a non-native field gagdet that is the result of a multiplication, n+1 should be divisible by 2^bpl, hence avoiding rounding.

Reply:
> I would still use 2^bpl since, at least for a non-native field gagdet that is the result of a multiplication, n+1 should be divisible by 2^bpl, hence avoiding rounding.

It depends if you change the conversion from NonNativeFieldMulResultGadget -> NonNativeFieldGadget
according to my comment below. In that case (n+1) will be a multiple o (2^bpl - 1).

Reply:
Uhm not sure, consider that n+1 is computed in `mul_without_prereduce_for_pseudomersenne` to be multiple of 2^bpl, it is not initialized from a NonNativeMulResultGadget with the function you referred to.

Reply:
Formula fixed in the code employing 2^bpl